File: src/handlers/bedrock.ts
import axios from 'axios';
import { BedrockModel, BedrockModelSearchResult } from '../types/bedrock.js';
import { PackageHandler } from '../types/index.js';

export class BedrockHandler implements PackageHandler {
  private modelsCache: BedrockModel[] | null = null;
  private lastFetchTime: number = 0;
  private readonly cacheTTL = 3600000; // 1 hour in milliseconds
  private readonly docsUrl = 'https://docs.aws.amazon.com/bedrock/latest/userguide/models-supported.html';

  constructor() {}

  /**
   * Fetches the latest Bedrock model information from AWS documentation
   */
  private async fetchModels(): Promise<BedrockModel[]> {
    // Check if we have a valid cache
    const now = Date.now();
    if (this.modelsCache && (now - this.lastFetchTime < this.cacheTTL)) {
      return this.modelsCache;
    }

    try {
      const response = await axios.get(this.docsUrl);
      const html = response.data;

      // Parse the HTML to extract the model information from the first table
      const models = this.parseModelsFromHtml(html);

      // Update cache
      this.modelsCache = models;
      this.lastFetchTime = now;

      return models;
    } catch (error) {
      console.error('Error fetching Bedrock models:', error);
      // If we have a cache, return it even if it's expired
      if (this.modelsCache) {
        return this.modelsCache;
      }
      throw error;
    }
  }

  /**
   * Parses the HTML to extract model information from the first table
   */
  private parseModelsFromHtml(html: string): BedrockModel[] {
    const models: BedrockModel[] = [];

    try {
      // Find the first table in the HTML
      // AWS docs typically use more complex table structure with classes
      const tableRegex = /<table[\s\S]*?>[\s\S]*?<\/table>/i;
      const tableMatch = html.match(tableRegex);

      if (!tableMatch) {
        console.error('No table found in HTML');
        return models;
      }

      const tableHtml = tableMatch[0];

      // Extract rows from the table (skip the header row)
      const rowRegex = /<tr[\s\S]*?>[\s\S]*?<\/tr>/gi;
      const rows = tableHtml.match(rowRegex);

      if (!rows || rows.length <= 1) {
        console.error('No rows found in table');
        return models;
      }

      // Skip the header row
      for (let i = 1; i < rows.length; i++) {
        const row = rows[i];

        // Extract cells from the row - AWS docs might use th or td
        const cellRegex = /<t[dh][\s\S]*?>[\s\S]*?<\/t[dh]>/gi;
        const cells = row.match(cellRegex);

        if (!cells || cells.length < 7) {
          console.error(`Invalid row format (cells: ${cells?.length}):`, row.substring(0, 100) + '...');
          continue;
        }

        // Extract text from cells
        const provider = this.extractTextFromCell(cells[0]);
        const modelName = this.extractTextFromCell(cells[1]);
        const modelId = this.extractTextFromCell(cells[2]);
        const regionsSupported = this.extractRegionsFromCell(cells[3]);
        const inputModalities = this.extractListFromCell(cells[4]);
        const outputModalities = this.extractListFromCell(cells[5]);
        const streamingSupported = this.extractTextFromCell(cells[6]).toLowerCase() === 'yes';

        // Only add if we have valid data
        if (modelName && modelId) {
          models.push({
            provider,
            modelName,
            modelId,
            regionsSupported,
            inputModalities,
            outputModalities,
            streamingSupported
          });
        }
      }

      console.log(`Successfully parsed ${models.length} models from HTML`);
    } catch (error) {
      console.error('Error parsing HTML:', error);
    }

    return models;
  }

  /**
   * Extracts text from a table cell
   */
  private extractTextFromCell(cell: string): string {
    // Remove HTML tags and trim whitespace
    return cell.replace(/<[^>]*>/g, '').trim();
  }

  /**
   * Extracts a list of regions from a cell
   */
  private extractRegionsFromCell(cell: string): string[] {
    const text = this.extractTextFromCell(cell);
    // Split by whitespace and filter out empty strings
    return text.split(/\s+/).filter(region => region.length > 0);
  }

  /**
   * Extracts a list from a cell (comma-separated values)
   */
  private extractListFromCell(cell: string): string[] {
    const text = this.extractTextFromCell(cell);
    // Split by comma and trim whitespace
    return text.split(',').map(item => item.trim()).filter(item => item.length > 0);
  }

  /**
   * Searches for Bedrock models based on query parameters
   */
  public async searchModels(query: string, provider?: string, region?: string): Promise<BedrockModelSearchResult> {
    const models = await this.fetchModels();

    // Normalize the query for case-insensitive search
    const normalizedQuery = query.toLowerCase().trim();
    const normalizedProvider = provider?.toLowerCase().trim();
    const normalizedRegion = region?.toLowerCase().trim();

    // Filter models based on query parameters
    const filteredModels = models.filter(model => {
      // Match query against model name or model ID (fuzzy search)
      const matchesQuery = !normalizedQuery ||
        this.fuzzyMatch(model.modelName.toLowerCase(), normalizedQuery) ||
        this.fuzzyMatch(model.modelId.toLowerCase(), normalizedQuery);

      // Match provider if specified (fuzzy search)
      const matchesProvider = !normalizedProvider ||
        this.fuzzyMatch(model.provider.toLowerCase(), normalizedProvider);

      // Match region if specified
      const matchesRegion = !normalizedRegion ||
        model.regionsSupported.some(r =>
          r.toLowerCase().includes(normalizedRegion) ||
          normalizedRegion.includes(r.toLowerCase())
        );

      return matchesQuery && matchesProvider && matchesRegion;
    });

    // Sort results by relevance if there's a query
    let sortedModels = filteredModels;
    if (normalizedQuery) {
      sortedModels = filteredModels.sort((a, b) => {
        // Exact matches first
        const aExactMatch =
          a.modelName.toLowerCase() === normalizedQuery ||
          a.modelId.toLowerCase() === normalizedQuery;
        const bExactMatch =
          b.modelName.toLowerCase() === normalizedQuery ||
          b.modelId.toLowerCase() === normalizedQuery;

        if (aExactMatch && !bExactMatch) return -1;
        if (!aExactMatch && bExactMatch) return 1;

        // Then sort by how early the match appears
        const aNameIndex = a.modelName.toLowerCase().indexOf(normalizedQuery);
        const bNameIndex = b.modelName.toLowerCase().indexOf(normalizedQuery);
        const aIdIndex = a.modelId.toLowerCase().indexOf(normalizedQuery);
        const bIdIndex = b.modelId.toLowerCase().indexOf(normalizedQuery);

        const aIndex = aNameIndex >= 0 ? aNameIndex : (aIdIndex >= 0 ? aIdIndex : Number.MAX_SAFE_INTEGER);
        const bIndex = bNameIndex >= 0 ? bNameIndex : (bIdIndex >= 0 ? bIdIndex : Number.MAX_SAFE_INTEGER);

        return aIndex - bIndex;
      });
    }

    return {
      models: sortedModels,
      totalCount: sortedModels.length
    };
  }

  /**
   * Performs a fuzzy match between a string and a query
   */
  private fuzzyMatch(str: string, query: string): boolean {
    // Simple implementation: check if all characters in query appear in order in str
    if (!query) return true;
    if (!str) return false;

    // Direct substring match
    if (str.includes(query)) return true;

    // Check for character-by-character fuzzy match
    let strIndex = 0;
    let queryIndex = 0;

    while (strIndex < str.length && queryIndex < query.length) {
      if (str[strIndex] === query[queryIndex]) {
        queryIndex++;
      }
      strIndex++;
    }

    return queryIndex === query.length;
  }

  /**
   * Lists all available Bedrock models
   */
  public async listModels(): Promise<BedrockModelSearchResult> {
    const models = await this.fetchModels();

    return {
      models,
      totalCount: models.length
    };
  }

  /**
   * Gets a specific Bedrock model by ID
   */
  public async getModelById(modelId: string): Promise<BedrockModel | null> {
    const models = await this.fetchModels();

    return models.find(model => model.modelId === modelId) || null;
  }

  /**
   * Gets the latest Claude Sonnet model
   * This returns the most recent version of Claude Sonnet, which is typically
   * the best model for coding tasks
   */
  public async getLatestClaudeSonnetModel(): Promise<BedrockModel | null> {
    const models = await this.fetchModels();

    // Filter for Claude Sonnet models
    const sonnetModels = models.filter(model =>
      model.provider.toLowerCase().includes('anthropic') &&
      model.modelName.toLowerCase().includes('claude') &&
      model.modelName.toLowerCase().includes('sonnet')
    );

    if (sonnetModels.length === 0) {
      return null;
    }

    // Sort by version number to find the latest
    // Extract version numbers and sort
    const sortedModels = sonnetModels.sort((a, b) => {
      // Extract version numbers (e.g., 3.5, 3.7)
      const aVersionMatch = a.modelName.match(/(\d+\.\d+)/);
      const bVersionMatch = b.modelName.match(/(\d+\.\d+)/);

      const aVersion = aVersionMatch ? parseFloat(aVersionMatch[1]) : 0;
      const bVersion = bVersionMatch ? parseFloat(bVersionMatch[1]) : 0;

      // First compare by version number (higher is better)
      if (aVersion !== bVersion) {
        return bVersion - aVersion;
      }

      // If same version number, check for "v2" or similar in the name
      const aHasV2 = a.modelName.toLowerCase().includes('v2');
      const bHasV2 = b.modelName.toLowerCase().includes('v2');

      if (aHasV2 && !bHasV2) return -1;
      if (!aHasV2 && bHasV2) return 1;

      // If still tied, prefer newer model IDs (they often have dates in them)
      return b.modelId.localeCompare(a.modelId);
    });

    return sortedModels[0];
  }

  /**
   * Implementation of the PackageHandler interface
   */
  public async getLatestVersion(args: any): Promise<{ content: { type: string; text: string }[] }> {
    try {
      let result: BedrockModelSearchResult | BedrockModel | null;

      if (args.action === 'search') {
        result = await this.searchModels(args.query || '', args.provider, args.region);
      } else if (args.action === 'get') {
        const model = await this.getModelById(args.modelId);
        result = {
          models: model ? [model] : [],
          totalCount: model ? 1 : 0
        };
      } else if (args.action === 'get_latest_claude_sonnet') {
        // Get the latest Claude Sonnet model
        const model = await this.getLatestClaudeSonnetModel();

        if (!model) {
          return {
            content: [
              {
                type: 'text',
                text: JSON.stringify({ error: 'No Claude Sonnet model found' }, null, 2)
              }
            ]
          };
        }

        // Return just the model, not wrapped in a search result
        result = model;
      } else {
        // Default to list all models
        result = await this.listModels();
      }

      return {
        content: [
          {
            type: 'text',
            text: JSON.stringify(result, null, 2)
          }
        ]
      };
    } catch (error) {
      console.error('Error in BedrockHandler:', error);
      return {
        content: [
          {
            type: 'text',
            text: `Error fetching Bedrock models: ${error instanceof Error ? error.message : String(error)}`
          }
        ]
      };
    }
  }
}

File: internal/handlers/bedrock.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"sync"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/sirupsen/logrus"
)

// BedrockHandler handles AWS Bedrock model checking
type BedrockHandler struct {
	client HTTPClient
	cache  *sync.Map
	logger *logrus.Logger
}

// NewBedrockHandler creates a new Bedrock handler
func NewBedrockHandler(logger *logrus.Logger, cache *sync.Map) *BedrockHandler {
	if cache == nil {
		cache = &sync.Map{}
	}
	return &BedrockHandler{
		client: DefaultHTTPClient,
		cache:  cache,
		logger: logger,
	}
}

// GetLatestVersion gets information about AWS Bedrock models
func (h *BedrockHandler) GetLatestVersion(ctx context.Context, args map[string]interface{}) (*mcp.CallToolResult, error) {
	h.logger.Info("Getting AWS Bedrock model information")

	// Parse action
	action := "list"
	if actionRaw, ok := args["action"].(string); ok && actionRaw != "" {
		action = actionRaw
	}

	// Handle different actions
	switch action {
	case "list":
		return h.listModels()
	case "search":
		return h.searchModels(args)
	case "get":
		return h.getModel(args)
	case "get_latest_claude_sonnet":
		return h.getLatestClaudeSonnet()
	default:
		return nil, fmt.Errorf("invalid action: %s", action)
	}
}

// listModels lists all available AWS Bedrock models
func (h *BedrockHandler) listModels() (*mcp.CallToolResult, error) {
	// In a real implementation, this would fetch data from AWS Bedrock API
	// For now, we'll return a static list of models
	models := []BedrockModel{
		{
			Provider:           "anthropic",
			ModelName:          "Claude 3 Opus",
			ModelID:            "anthropic.claude-3-opus-20240229-v1:0",
			RegionsSupported:   []string{"us-east-1", "us-west-2", "eu-central-1"},
			InputModalities:    []string{"text", "image"},
			OutputModalities:   []string{"text"},
			StreamingSupported: true,
		},
		{
			Provider:           "anthropic",
			ModelName:          "Claude 3 Sonnet",
			ModelID:            "anthropic.claude-3-sonnet-20240229-v1:0",
			RegionsSupported:   []string{"us-east-1", "us-west-2", "eu-central-1"},
			InputModalities:    []string{"text", "image"},
			OutputModalities:   []string{"text"},
			StreamingSupported: true,
		},
		{
			Provider:           "anthropic",
			ModelName:          "Claude 3 Haiku",
			ModelID:            "anthropic.claude-3-haiku-20240307-v1:0",
			RegionsSupported:   []string{"us-east-1", "us-west-2", "eu-central-1"},
			InputModalities:    []string{"text", "image"},
			OutputModalities:   []string{"text"},
			StreamingSupported: true,
		},
		{
			Provider:           "amazon",
			ModelName:          "Titan Text G1 - Express",
			ModelID:            "amazon.titan-text-express-v1",
			RegionsSupported:   []string{"us-east-1", "us-west-2"},
			InputModalities:    []string{"text"},
			OutputModalities:   []string{"text"},
			StreamingSupported: true,
		},
		{
			Provider:           "amazon",
			ModelName:          "Titan Image Generator G1",
			ModelID:            "amazon.titan-image-generator-v1",
			RegionsSupported:   []string{"us-east-1", "us-west-2"},
			InputModalities:    []string{"text"},
			OutputModalities:   []string{"image"},
			StreamingSupported: false,
		},
		{
			Provider:           "cohere",
			ModelName:          "Command",
			ModelID:            "cohere.command-text-v14",
			RegionsSupported:   []string{"us-east-1", "us-west-2"},
			InputModalities:    []string{"text"},
			OutputModalities:   []string{"text"},
			StreamingSupported: true,
		},
		{
			Provider:           "meta",
			ModelName:          "Llama 2 Chat 13B",
			ModelID:            "meta.llama2-13b-chat-v1",
			RegionsSupported:   []string{"us-east-1", "us-west-2"},
			InputModalities:    []string{"text"},
			OutputModalities:   []string{"text"},
			StreamingSupported: true,
		},
		{
			Provider:           "meta",
			ModelName:          "Llama 2 Chat 70B",
			ModelID:            "meta.llama2-70b-chat-v1",
			RegionsSupported:   []string{"us-east-1", "us-west-2"},
			InputModalities:    []string{"text"},
			OutputModalities:   []string{"text"},
			StreamingSupported: true,
		},
		{
			Provider:           "stability",
			ModelName:          "Stable Diffusion XL 1.0",
			ModelID:            "stability.stable-diffusion-xl-v1",
			RegionsSupported:   []string{"us-east-1", "us-west-2"},
			InputModalities:    []string{"text"},
			OutputModalities:   []string{"image"},
			StreamingSupported: false,
		},
	}

	// Sort models by provider and name
	sort.Slice(models, func(i, j int) bool {
		if models[i].Provider != models[j].Provider {
			return models[i].Provider < models[j].Provider
		}
		return models[i].ModelName < models[j].ModelName
	})

	result := BedrockModelSearchResult{
		Models:     models,
		TotalCount: len(models),
	}

	return NewToolResultJSON(result)
}

// searchModels searches for AWS Bedrock models
func (h *BedrockHandler) searchModels(args map[string]interface{}) (*mcp.CallToolResult, error) {
	// Get all models
	result, err := h.listModels()
	if err != nil {
		return nil, err
	}

	// Convert result to JSON string
	resultJSON, err := json.Marshal(result)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal result: %w", err)
	}

	// Parse result
	var data map[string]interface{}
	if err := json.Unmarshal(resultJSON, &data); err != nil {
		return nil, fmt.Errorf("failed to parse model data: %w", err)
	}

	// Get models
	modelsRaw, ok := data["models"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid model data format")
	}

	// Parse query
	query := ""
	if queryRaw, ok := args["query"].(string); ok {
		query = strings.ToLower(queryRaw)
	}

	// Parse provider
	provider := ""
	if providerRaw, ok := args["provider"].(string); ok {
		provider = strings.ToLower(providerRaw)
	}

	// Parse region
	region := ""
	if regionRaw, ok := args["region"].(string); ok {
		region = strings.ToLower(regionRaw)
	}

	// Filter models
	var filteredModels []BedrockModel
	for _, modelRaw := range modelsRaw {
		modelMap, ok := modelRaw.(map[string]interface{})
		if !ok {
			continue
		}

		// Convert to BedrockModel
		var model BedrockModel
		modelJSON, err := json.Marshal(modelMap)
		if err != nil {
			continue
		}
		if err := json.Unmarshal(modelJSON, &model); err != nil {
			continue
		}

		// Apply filters
		if query != "" {
			nameMatch := strings.Contains(strings.ToLower(model.ModelName), query)
			idMatch := strings.Contains(strings.ToLower(model.ModelID), query)
			providerMatch := strings.Contains(strings.ToLower(model.Provider), query)
			if !nameMatch && !idMatch && !providerMatch {
				continue
			}
		}

		if provider != "" && !strings.Contains(strings.ToLower(model.Provider), provider) {
			continue
		}

		if region != "" {
			var regionMatch bool
			for _, r := range model.RegionsSupported {
				if strings.Contains(strings.ToLower(r), region) {
					regionMatch = true
					break
				}
			}
			if !regionMatch {
				continue
			}
		}

		filteredModels = append(filteredModels, model)
	}

	// Sort models by provider and name
	sort.Slice(filteredModels, func(i, j int) bool {
		if filteredModels[i].Provider != filteredModels[j].Provider {
			return filteredModels[i].Provider < filteredModels[j].Provider
		}
		return filteredModels[i].ModelName < filteredModels[j].ModelName
	})

	searchResult := BedrockModelSearchResult{
		Models:     filteredModels,
		TotalCount: len(filteredModels),
	}

	return NewToolResultJSON(searchResult)
}

// getModel gets a specific AWS Bedrock model
func (h *BedrockHandler) getModel(args map[string]interface{}) (*mcp.CallToolResult, error) {
	// Parse model ID
	modelID, ok := args["modelId"].(string)
	if !ok || modelID == "" {
		return nil, fmt.Errorf("missing required parameter: modelId")
	}

	// Get all models
	result, err := h.listModels()
	if err != nil {
		return nil, err
	}

	// Convert result to JSON string
	resultJSON, err := json.Marshal(result)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal result: %w", err)
	}

	// Parse result
	var data map[string]interface{}
	if err := json.Unmarshal(resultJSON, &data); err != nil {
		return nil, fmt.Errorf("failed to parse model data: %w", err)
	}

	// Get models
	modelsRaw, ok := data["models"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid model data format")
	}

	// Find model
	for _, modelRaw := range modelsRaw {
		modelMap, ok := modelRaw.(map[string]interface{})
		if !ok {
			continue
		}

		// Check model ID
		if id, ok := modelMap["modelId"].(string); ok && id == modelID {
			return NewToolResultJSON(modelMap)
		}
	}

	return nil, fmt.Errorf("model not found: %s", modelID)
}

// getLatestClaudeSonnet gets the latest Claude Sonnet model
func (h *BedrockHandler) getLatestClaudeSonnet() (*mcp.CallToolResult, error) {
	// Get all models
	result, err := h.listModels()
	if err != nil {
		return nil, err
	}

	// Convert result to JSON string
	resultJSON, err := json.Marshal(result)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal result: %w", err)
	}

	// Parse result
	var data map[string]interface{}
	if err := json.Unmarshal(resultJSON, &data); err != nil {
		return nil, fmt.Errorf("failed to parse model data: %w", err)
	}

	// Get models
	modelsRaw, ok := data["models"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid model data format")
	}

	// Find Claude Sonnet model
	for _, modelRaw := range modelsRaw {
		modelMap, ok := modelRaw.(map[string]interface{})
		if !ok {
			continue
		}

		// Convert to BedrockModel
		var model BedrockModel
		modelJSON, err := json.Marshal(modelMap)
		if err != nil {
			continue
		}
		if err := json.Unmarshal(modelJSON, &model); err != nil {
			continue
		}

		// Check if it's Claude Sonnet
		if model.Provider == "anthropic" && strings.Contains(model.ModelName, "Sonnet") {
			return NewToolResultJSON(model)
		}
	}

	return nil, fmt.Errorf("Claude Sonnet model not found")
}
